src/libsystemd/sd-journal/journal-file.c

```diff
@@ -318,13 +318,17 @@ static int journal_file_init_header(JournalFile *f, JournalFileFlags file_flags,
         seal = FLAGS_SET(file_flags, JOURNAL_SEAL) && journal_file_fss_load(f) >= 0;
 #endif
 
-        r = getenv_bool("SYSTEMD_JOURNAL_COMPACT");
-        if (r < 0) {
-                if (r != -ENXIO)
-                        log_debug_errno(r, "Failed to parse $SYSTEMD_JOURNAL_COMPACT environment variable, ignoring: %m");
-                compact = true;
-        } else
-                compact = r;
+        if (FLAGS_SET(file_flags, JOURNAL_MULTI_BOOT))
+                compact = false;
```

Comment:
maybe add explicit comment here: `/* the compact mode does not support multiple different boot ids per journal file, hence disable compat mode if that's requested */